### PR TITLE
[symbolic shapes] if symbol not in var_ranges default to unknown range

### PR DESCRIPTION
**Pull request by ColinPeppler**
Purpose of this PR is to get around this error: https://github.com/pytorch/pytorch/issues/127677

Differential Revision: D58048558

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/index_propagation.py [hunk 1]
```diff
@@ -333,16 +335,20 @@ def wrap_expr(expr):
...
+                expr = wrap_expr(expr)
+                if generate_assert(check):
+                    self.fallback(
+                        "check_bounds",
+                        (expr, size),
+                        dict(lower=not can_prove_lower, upper=not can_prove_upper),
+                    )
+                return expr
+            except KeyError as e:
+                log.warn(str(e))
```
Comment (Skylion007) [inline torch/_inductor/index_propagation.py:352]: Use 'log.warning("", exc_info=True)'

**Review by lezcano — CHANGES_REQUESTED** [review]
I would really like to understand https://github.com/pytorch/pytorch/issues/127677#issuecomment-2143477034, as this is just a hint that there's another bug lurking underneath.

**Comment by ColinPeppler** [comment]
@lezcano 
I'd love to take a look, but I also need to urgently resolve an error happening internally. Can we slap a bandaid now and I'll return to this ASAP?

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -4411,7 +4411,11 @@ def _maybe_evaluate_static(
                 # Skip var_ranges logic for SingletonInt which is only used
                 # for jagged layout NestedTensors today
                 continue
-            vr = var_ranges[k]
+            try:
+                vr = self.var_to_range[k]
+            except KeyError:
+                log.warning(f"{k} is not in var_to_range, defaulting to unknown range.")
+                vr = ValueRanges(-sys.maxsize, sys.maxsize)
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:4418]: Use `self._default_unspecified_value_range()` then, as this one has an off-by-one error.

**Review by lezcano — COMMENTED** [review]
Sure, let's land this as a hotfix and leave the other issue open.

**Comment by ColinPeppler** [comment]
sounds good, assigned myself to the issue

**Review by lezcano — APPROVED** [review]

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -4411,7 +4411,11 @@ def _maybe_evaluate_static(
                 # Skip var_ranges logic for SingletonInt which is only used
                 # for jagged layout NestedTensors today
                 continue
-            vr = var_ranges[k]
+            try:
+                vr = self.var_to_range[k]
+            except KeyError:
+                log.warning("{} is not in var_to_range, defaulting to unknown range.".format(k))
```
Comment (Skylion007) [inline torch/fx/experimental/symbolic_shapes.py:4417]:
```suggestion
                log.warning("%s is not in var_to_range, defaulting to unknown range.", k)
```

Comment (ColinPeppler) [reply torch/fx/experimental/symbolic_shapes.py]: thx! was looking for the proper way to do that

**Comment by bhack** [comment]
With recently nighlty I've just got many of this warning.
```python
W0605 11:48:51.002000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] q0 is not in var_ranges, defaulting to unknown range.
W0605 11:48:51.002000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] q1 is not in var_ranges, defaulting to unknown range.
W0605 11:48:51.013000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] q0 is not in var_ranges, defaulting to unknown range.
W0605 11:48:51.013000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] q1 is not in var_ranges, defaulting to unknown range.
W0605 11:48:51.324000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] z1 is not in var_ranges, defaulting to unknown range.
W0605 11:48:51.324000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] z0 is not in var_ranges, defaulting to unknown range.
W0605 11:48:51.334000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] z1 is not in var_ranges, defaulting to unknown range.
W0605 11:48:51.334000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] z0 is not in var_ranges, defaulting to unknown range.
W0605 11:48:52.669000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] q0 is not in var_ranges, defaulting to unknown range.
W0605 11:48:52.800000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] z0 is not in var_ranges, defaulting to unknown range.
W0605 11:48:54.093000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] q0 is not in var_ranges, defaulting to unknown range.
W0605 11:48:54.093000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] q1 is not in var_ranges, defaulting to unknown range.
W0605 11:48:54.152000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] z1 is not in var_ranges, defaulting to unknown range.
W0605 11:48:54.152000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] z0 is not in var_ranges, defaulting to unknown range.
W0605 11:48:54.734000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] q0 is not in var_ranges, defaulting to unknown range.
W0605 11:48:54.869000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] z0 is not in var_ranges, defaulting to unknown range.
W0605 11:48:56.124000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] q0 is not in var_ranges, defaulting to unknown range.
W0605 11:48:56.124000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] q1 is not in var_ranges, defaulting to unknown range.
W0605 11:48:56.189000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] z1 is not in var_ranges, defaulting to unknown range.
W0605 11:48:56.189000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] z0 is not in var_ranges, defaulting to unknown range.
W0605 11:48:57.112000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] q0 is not in var_ranges, defaulting to unknown range.
W0605 11:48:57.240000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] z0 is not in var_ranges, defaulting to unknown range.
W0605 11:49:03.713000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] y1 is not in var_ranges, defaulting to unknown range.
W0605 11:49:03.713000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] y0 is not in var_ranges, defaulting to unknown range.
W0605 11:49:03.725000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] y1 is not in var_ranges, defaulting to unknown range.
W0605 11:49:03.725000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] y0 is not in var_ranges, defaulting to unknown range.
W0605 11:49:04.647000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] x0 is not in var_ranges, defaulting to unknown range.
W0605 11:49:05.445000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] x2 is not in var_ranges, defaulting to unknown range.
W0605 11:49:05.445000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] x1 is not in var_ranges, defaulting to unknown range.
W0605 11:49:05.923000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] x0 is not in var_ranges, defaulting to unknown range.
W0605 11:49:06.703000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] x2 is not in var_ranges, defaulting to unknown range.
W0605 11:49:06.703000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] x1 is not in var_ranges, defaulting to unknown range.
W0605 11:49:07.186000 134494347568960 torch/fx/experimental/symbolic_shapes.py:4417] [0/0] x0 is not in var_ranges, defaulting to unknown range.
````

Is there any user action item on these?

**Comment by lezcano** [comment]
@ColinPeppler will look into how to properly fix https://github.com/pytorch/pytorch/issues/127677

**Comment by ColinPeppler** [comment]
@bhack those warnings would've led to failures. And if you have a minimal repro, pls share in #127677 then I can check with the fix.

**Comment by bhack** [comment]
@ColinPeppler The code is quite large to isolate. I have the full env in another ticket if you want.
If the full env is too much for you do you need somethin from https://pytorch.org/docs/stable/torch.compiler_troubleshooting.html ?

**Comment by lezcano** [comment]
Could you try to isolate a smaller repro by putting breakpoints on this warnign and seeing the code that calls it @bhack and doing some manual binary search in the resulting fx graph?

**Comment by bhack** [comment]
@lezcano A first isolation it is at https://github.com/pytorch/pytorch/issues/127677#issuecomment-2152532190

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -4411,7 +4411,11 @@ def _maybe_evaluate_static(
                 # Skip var_ranges logic for SingletonInt which is only used
                 # for jagged layout NestedTensors today
                 continue
-            vr = var_ranges[k]
+            try:
+                vr = var_ranges[k]
+            except KeyError:
+                log.warning("%s is not in var_ranges, defaulting to unknown range.", k)
```
Comment (jgong5) [inline torch/fx/experimental/symbolic_shapes.py:4417]: May I know why it is a warning here? Would it always warn on unbacked symbols? The cpp template uses some unbacked symbols to build IR nodes for codegen which would always get warning here.
https://github.com/pytorch/pytorch/blob/695502ca653d25be23be61a909587f110067e859/torch/_inductor/codegen/cpp_gemm_template.py#L110

**Comment by lezcano** [comment]
@ColinPeppler this is annoying too many people. Can you send a patch that removes the warning, or send a fix altogether?

**Comment by ColinPeppler** [comment]
sorry if I can't post a fix by end-of-today, I will remove the warning.

**Comment by bhack** [comment]
Is there a fix?

**Pull request closed without merging**